### PR TITLE
Support to verify authorized party (azp) session token claim

### PR DESCRIPTION
**Pull request by chanioxaris**
## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/backend-core`
- [x] `@clerk/clerk-sdk-node`
- [x] `@clerk/edge`
- [ ] `tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

### CODE REVIEW

File: packages/backend-core/src/Base.ts [hunk 1]
```diff
@@ -88,9 +90,10 @@ export class Base {
    * The public key will be supplied in the form of CryptoKey or will be loaded from the CLERK_JWT_KEY environment variable.
    *
    * @param {string} token
+   * @param authorizedParties
    * @return {Promise<JWTPayload>} claims
    */
-  verifySessionToken = async (token: string): Promise<JWTPayload> => {
+  verifySessionToken = async (token: string, authorizedParties?: string[]): Promise<JWTPayload> => {
```
Comment (SokratisVidros) [inline packages/backend-core/src/Base.ts:96]: 🙃 Let's make the `authorizedParties` an object as in the future we will need to add more parameters from the user.

File: packages/sdk-node/src/Clerk.ts [hunk 2]
```diff
@@ -169,8 +170,14 @@ export default class Clerk extends ClerkBackendAPI {
       throw new Error('Malformed token');
     }
 
-    if (!(verified.iss?.lastIndexOf('https://clerk.', 0) === 0)) {
-      throw new Error(`Invalid issuer: ${verified.iss}`);
+    if (!verified.iss || !verified.iss.startsWith('https://clerk')) {
+      throw new Error(`Issuer is invalid: ${verified.iss}`);
+    }
+
+    if (verified.azp && authorizedParties && authorizedParties.length !== 0) {
```
Comment (SokratisVidros) [inline packages/sdk-node/src/Clerk.ts:177]:
```suggestion
    if (verified.azp && authorizedParties && authorizedParties.length > 0) {
```

**Review by SokratisVidros — CHANGES_REQUESTED** [review]
Nicely done @chanioxaris. Let's add a unit tests for the check as it's a crucial functionality

**Comment by chanioxaris** [comment]
All comments have been addressed. Also unit tests have been added with multiple cases to cover the new functionality

File: packages/backend-core/src/Base.ts [hunk 3]
```diff
@@ -88,9 +94,10 @@ export class Base {
    * The public key will be supplied in the form of CryptoKey or will be loaded from the CLERK_JWT_KEY environment variable.
    *
    * @param {string} token
+   * @param {string[]} authorizedParties
```
Comment (igneel64) [inline packages/backend-core/src/Base.ts:97]: 🙃 
I think this should actually be {VerifySessionTokenOptions} verifySessionTokenOptions

File: packages/backend-core/src/Base.ts [hunk 4]
```diff
@@ -88,9 +94,10 @@ export class Base {
...
    * @return {Promise<JWTPayload>} claims
    */
-  verifySessionToken = async (token: string): Promise<JWTPayload> => {
+  verifySessionToken = async (token: string, { authorizedParties }: VerifySessionTokenOptions = { authorizedParties: []}): Promise<JWTPayload> => {
```
Comment (igneel64) [inline packages/backend-core/src/Base.ts:100]: 🙃 
For now we only have one option but can we create an object called `verifySessionTokenDefaultOptions` and use it here ?

File: packages/backend-core/src/util/jwt.ts [hunk 5]
```diff
@@ -17,8 +17,14 @@ export function isExpired(
   }
 }
 
-export function checkClaims(claims: JWTPayload) {
+export function checkClaims(claims: JWTPayload, authorizedParties?: string[]) {
   if (!claims.iss || !claims.iss.startsWith('https://clerk')) {
     throw new Error(`Issuer is invalid: ${claims.iss}`);
   }
+
+  if (claims.azp && authorizedParties && authorizedParties.length > 0) {
```
Comment (igneel64) [inline packages/backend-core/src/util/jwt.ts:25]: 🙃 
Except if we want to be really strict, could we maybe reduce the check in this:
```ts
  if (!authorizedParties?.includes(claims.azp as string)) {
    throw new Error(`Authorized party is invalid: ${claims.azp}`);
  }
```

File: packages/backend-core/src/util/jwt.ts [hunk 6]
```diff
@@ -17,8 +17,14 @@ export function isExpired(
...
+    if (!authorizedParties.includes(claims.azp as string)) {
+      throw new Error(`Authorized party is invalid: ${claims.azp}`);
```
Comment (igneel64) [inline packages/backend-core/src/util/jwt.ts:27]: 🙃 
If our JWTs always contain azp, should we add this on the `JWTPayload` interface ? 
Same with `iss` above.

**Comment by igneel64** [comment]
@chanioxaris 
Good one! If you can also update the Readme for this option, it would be nice!

File: packages/backend-core/src/Base.ts (hunk 3 above)
Comment (chanioxaris) [reply packages/backend-core/src/Base.ts]: Fixed in [02cd6d8](https://github.com/clerkinc/javascript/pull/15/commits/02cd6d819029c3c66f8e308028e81c5731841cd8)

**Comment by chanioxaris** [comment]
> @chanioxaris Good one! If you can also update the Readme for this option, it would be nice!

@igneel64 README has also been updated [02cd6d8](https://github.com/clerkinc/javascript/pull/15/commits/02cd6d819029c3c66f8e308028e81c5731841cd8)

**Review by igneel64 — APPROVED** [review]

**Review by SokratisVidros — APPROVED** [review]

**Pull request merged**